### DISCUSSION TITLE
Migrate to buttercup for tests?

### DISCUSSION THREAD

**Discussion by jinnovation** [kubernetes-el/kubernetes-el#224] [Ideas]
@noorul, what are your thoughts about gradually shifting our testing over to use [buttercup](https://github.com/jorgenschaefer/emacs-buttercup)? 

There are several advantages over `ert`, which we currently use, including but not limited to the following:
- Native mocking/spying support, which `ert` does not have;
- Support for straightforward sharing of setup/teardown logic between multiple related test cases, e.g. via "before/after-each/all" blocks, which `ert` does not;
- User-friendly assertion constructs, e.g. `:to-equal` for equality or `:to-throw` for erroring, as well as ability to define custom ones (might be useful for temp-buffer content assertion for AST evaluation);
- In use by several high-profile Emacs packages, most notably [Projectile](https://github.com/bbatsov/projectile) -- good vote of confidence in buttercup's usage, as well as good opportunities for inspiration and learning from other packages.

I personally have used buttercup in other packages of mine, e.g. [bazel-transient](https://github.com/jinnovation/bazel-transient), and have found it to be a great experience that's consistent with what I expect from "modern" unit-testing frameworks in other languages, e.g. Python or Go.

Considering, as well, the relative lack of test coverage in `kubernetes-el` currently, anything that we can do to facilitate ease of test writing would be IMO beneficial -- another argument in favor of this switch.

I don't envision this transition needing to be an immediate, all-or-nothing endeavor; we could gradually convert tests over to buttercup as we see fit.

**Comment by jinnovation** [comment]
As a concrete example of the benefits that Buttercup stands to provide us, consider [the following test](https://github.com/kubernetes-el/kubernetes-el/blob/master/test/kubernetes-ast-test.el#L38-L42):
```emacs-lisp
(ert-deftest kubernetes-ast-test--inserts-strings ()
  (let ((ast '(line "foo")))
    (with-temp-buffer
      (kubernetes-ast-eval ast)
      (should (equal "foo\n" (buffer-string))))))
```
This pattern—evaluate AST, insert results into temp buffer, and compare buffer contents to the expected string value—is repeated for every AST evaluation test that doesn't cover an error case. This represents a substantial amount of boilerplate that impedes both the writing of tests as well as the reading.

With buttercup, we could abstract away those irrelevant mechanics of the test, encapsulating them into a [matcher](https://github.com/jorgenschaefer/emacs-buttercup/blob/master/docs/writing-tests.md#matchers)—say, `:to-render-to`. Reference implementation like so:
```emacs-lisp
(buttercup-define-matcher :to-render-to (ast s-expect)
  (cl-destructuring-bind
      ((ast-expr . ast) (s-expect-expr . s-expect))
      (mapcar #'buttercup--expr-and-value (list ast s-expect))
    (with-temp-buffer
      (kubernetes-ast-eval ast)
      (buttercup--test-expectation
          (funcall 'equal (buffer-string) s-expect)
        :expect-match-phrase
        (format "Expected %s to render to \"%s\", but it rendered to \"%s\"" ast-expr s-expect (buffer-string))
        :expect-mismatch-phrase "Expected %s not to render to \"%s\", but it did."))))
```

This allows us to reduce the test itself very tightly to input and expected output, like so:
```emacs-lisp
(describe "AST rendering"
  (describe "lines"
    (it "inserts strings"
      (expect '(line "foo") :to-render-to "foo\n")))
```

This makes it *super* easy for contributors to add new test cases—again, without the overhead of setting up the mechanics of those tests themselves. Also aids readability of existing tests significantly, which can help in identifying untested cases.

**Comment by noorul** [comment]
The only concern is future contributors learning curve if someone is coming with ERT knowledge. The last commit in that project is on Jan 6. Is it actively maintained?

**Reply by jinnovation** [reply]
> The only concern is future contributors learning curve if someone is coming with ERT knowledge. 

We have the same risk w/ ERT for contributors (like myself) who previously might only have had experience writing tests w/ Buttercup. :P Besides, Buttercup is arguably a fairly established part of the Emacs development ecosystem, so I think the knowledge overhead would be justifiable.

> The last commit in that project is on Jan 6. Is it actively maintained?

This is a fair concern. That said, in this case, I suspect that the relative lack of activity speaks more to the relative maturity/stability of the project. Not to mention there seems to be maintainer activity on at least one recently opened issue.

If anything, buttercup might be more actively worked on and be more amenable to external contributions than ERT itself. Taking mock/stub functionality as an example, [ERT documentation](https://www.gnu.org/software/emacs/manual/html_mono/ert.html#Mocks-and-Stubs) refers you to [an EmacsWiki-hosted "package" that presumably hasn't been touched since 2012](https://www.emacswiki.org/emacs/el-mock.el). And ERT itself possibly doesn't receive a whole lot of recent attention either, judging by [changelogs for Emacs 28](https://github.com/emacs-mirror/emacs/blob/master/etc/NEWS.28) etc.—again, presumably due in part to the maturity of the package.

**Comment by noorul** [comment]
Having said that I have no objection to the move.

**Reply by jinnovation** [reply]
Sounds good. This is also a fairly [reversible decision](https://www.cnbc.com/2018/11/19/jeff-bezos-simple-strategy-for-answering-amazons-hardest-questions--.html), so if this turns out to be a bad idea we could always switch back to ERT or something else.

I'll put up an inaugural PR sometime converting a handful of our tests to Buttercup, so that we can get a feel for how things might look concretely. At the same time, will also make sure that CI, code coverage, etc. are accounted for.

**Reply by noorul** [reply]
Yes, please!

**Reply by jinnovation** [reply]
See #225 for preliminary example.
